internal_ws/ykcompile/src/lib.rs

```diff
@@ -288,7 +290,7 @@ pub extern "sysv64" fn bh_new_vec() -> *mut Vec<FrameInfo> {
 
 /// Pushes a new `FrameInfo` instance onto the vector behind the pointer `vptr`. The `FrameInfo`
 /// instance is creates from a pointer to a symbol name and its length, a basic block index and a
-/// pointer to some allocated memory.
+/// pointer to some allocated memory. XXX need to document that `vptr` is consumed
```

Comment:
In this case `vptr` is not consumed. Should I point that out in the docs, or is it implied if it's not specifically pointed out?

Reply:
I think it would be excessive to point out every time something is not consumed. Then again, the raw pointers make it hard to know.

Perhaps it's OK in this scenario.

Reply:
> In this case vptr is not consumed.

But surely it is at line 306?

```rust
let mut v = unsafe { Box::from_raw(vptr) };
```

Reply:
Yes, but it's then "deconsumed" (regurgitated?) in 308. So in a way it's consumed, but also the pointer is still valid when the function finishes (and is indeed used in later calls to `bh_push_vec`).

Reply:
But now that I think about it, I'm not sure Rust guarantees that the pointer will be the same after calling `Box::into_raw`.

Reply:
It should then use `let mut v = unsafe { &mut *vptr };`.

Reply:
Will use that then. But won't we still be in trouble if the vector is resized and needs to be moved in memory, invalidating the pointer?

Reply:
This is a pointer to a vector. This vector contains a pointer to the allocated memory.

Reply:
Ah, yes of course! Thanks.

Reply:
There's no notion of "deconsume" unfortunately: there's a very good chance this code is UB, even with @bjorn3's suggestion. The question is: where is this called from?

Reply:
It's called from the compiled trace during a guard failure: https://github.com/softdevteam/yk/blob/master/internal_ws/ykcompile/src/lib.rs#L1479

Reply:
With my suggestion nothing is consumed. Only borrowed.

Reply:
Just to give some more information: We create a new vector with `bh_new_vec`, which is then boxed and the pointer stored in a register. We then call `bh_push_vec` one or more times, passing that pointer in as an argument (with @bjorn3 suggesting we then only mutably borrow the pointer to push elements onto the vec). Finally, we pass the vector pointer into `invoke_sinterp`, where it is consumed.

Reply:
The full code is:

```rust
    let mut v = unsafe { Box::from_raw(vptr) };
    v.push(fi);
    Box::into_raw(v);
```

If something else has a mutable reference to that at the same time, we're in UB AFAICS, because we're calling into the vector.

Reply:
With my suggestion that becomes

```rust
let mut v = unsafe { &mut *vptr };
v.push(fi);
```

This doesn't consume anything.

Reply:
Is "that" the vector itself? Can we downgrade the mutability after the push?

Reply:
`vptr` is a pointer to the vector, so `v` is a mutable reference to the vector.

Reply:
The problem is that if someone else has a mutable reference to the vector and calls this function, we've now got two `&mut` references to the same thing. That is likely (though with stacked borrows there are some exceptions) to be UB. In other words, there *must* be some documented constraints on `vptr` that the caller has to respect.

Reply:
Isn't the fact that `vptr` is a `mut *` already telling the caller that the pointer is likely to be mutated and thus this function shouldn't be called in parallel. Though I'm happy to add a warning to the doc string.

Reply:
> Isn't the fact that vptr is a mut * already telling the caller that the pointer is likely to be mutated and thus this function shouldn't be called in parallel

Given that some other functions consume the pointer (and thus the `mut *` is quasi-irrelevant) the answer is clearly "no" :) In other words, if every function consumed the pointers, that would be easy; if none of them consumed the pointers that would be a bit more difficult, but at least consistent; but what we've got is a mix.

Basically, raw pointers in Rust are incredibly dangerous -- much, much, more dangerous than in C -- so we have to treat them like unexploded bombs. It's difficult for us to be too cautious!

Let me ask the dangerous question: at the point we get a pointer to this Vec, can we guarantee with 100% confidence that none else holds a `&` or `&mut` reference to it? [It's totally fine if they have a `const *` or `mut *` reference at the same time, but `&` or `&mut` are strictly verboten.]

Reply:
> at the point we get a pointer to this Vec, can we guarantee with 100% confidence that none else holds a & or &mut reference to it?

With the current code? Yes. The pointer to this vector is only ever passed into two functions: `bh_vec_push` and `invoke_sinterp`, the former is always called before the latter, and the latter consumes the pointer after which it's not used any more (see also my summary comment of this code above).

Reply:
> The pointer to this vector is only ever passed into two functions

But, just to check, does any other code hold a `&`/`&mut` reference to this vec? *If* we convert it once with `Box::into_raw` and this is the only code which converts it back to a `&`/`&mut`, we're safe.

Reply:
No, the vector is created in `bh_new_vec` and then immediately boxed and returned to the assembler code, which only passes it to `bh_vec_push` and `invoke_sinterp`. So at the moment there is no `&`/`&mut` reference, and with @bjorn3's suggestion there will be one.

Reply:
OK, let's do it!

Reply:
Fixed in 2c47787fbaaf271fdb2fef39104bbf681eee6454.